File: src/enemies.c
#include "enemies.h"

// Enemies are initially inactive unless they are right next to the player
void ai_init(){
  for(int i=0;s.scene[i].type != '\0';i++){
    if((s.scene[i].type == 'e' || s.scene[i].type == '&') && s.scene[i].x >  s.scene[s.pli].x + (16*8)){
      s.scene[i].active = false;
    }
  }
}

void ai_killAt(int i){
  ai_kill(&s.scene[i]);
}

// Makes the enemy nonphysical and shorter / squashed.
void ai_kill(obj* e){
  au_play(SND_shot);
  (*e).physical = false;
  (*e).vy = 8;
}

void ai_update(){
  for(int i=0;s.scene[i].type != '\0';i++){
    if((s.scene[i].type == 'e' || s.scene[i].type == '&') && s.scene[i].x < s.scene[s.pli].x + (16*8)+16){
      ai_activate(s.scene[i].x, s.scene[i].y);
    }
  }
}

// Activates enemies in groups
void ai_activate(int x, int y){
  for(int i=0;s.scene[i].type != '\0'; i++){
    if(s.scene[i].x == x && s.scene[i].y == y && s.scene[i].active == false){
      s.scene[i].active = true;
      ai_activate(x-2*16, y);
      ai_activate(x-1*16, y);
      ai_activate(x+1*16, y);
      ai_activate(x+2*16, y);
    }
  }
}

File: src/enemies.zig
const c = @import("c.zig");

extern var s: c.state;

export fn ai_init() void {
    var i: usize = 0;
    while (s.scene[i].type != '\x00') : (i += 1) {
        if ((s.scene[i].type == 'e' or s.scene[i].type == '&') and s.scene[i].x > s.scene[@intCast(s.pli)].x + (16 * 8)) {
            s.scene[i].active = false;
        }
    }
}

export fn ai_killAt(i: c_int) void {
    ai_kill(&s.scene[@intCast(i)]);
}

export fn ai_kill(e: [*c]c.obj) void {
    c.au_play(c.SND_shot);
    e.*.physical = false;
    e.*.vy = 8;
}

export fn ai_update() void {
    var i: usize = 0;
    while (s.scene[i].type != '\x00') : (i += 1) {
        if ((s.scene[i].type == 'e' or s.scene[i].type == '&') and s.scene[i].x < s.scene[@intCast(s.pli)].x + (16 * 8) + 16) {
            ai_activate(s.scene[i].x, s.scene[i].y);
        }
    }
}

export fn ai_activate(x: c_int, y: c_int) void {
    var i: usize = 0;
    while (s.scene[i].type != '\x00') : (i += 1) {
        if (s.scene[i].x == x and s.scene[i].y == y and s.scene[i].active == false) {
            s.scene[i].active = true;
            ai_activate(x - 2 * 16, y);
            ai_activate(x - 1 * 16, y);
            ai_activate(x + 1 * 16, y);
            ai_activate(x + 2 * 16, y);
        }
    }
}
